### ISSUE TITLE
Allow "dune coq top" to step through unsaved files

### ISSUE THREAD

**Issue by rlepigre** [ocaml/dune#7670]
## Desired Behavior

(This is a follow up to https://github.com/ocaml/dune/issues/7355, which was not completely addressed.)

If a file `file.v` that has not yet be written to disk would be part of a `coq.theory` stanza, then `dune coq top file.v` should not fail, and allow you to step through the file.

## Current behavior

The command currently fails with `Error: Cannot find file: file.v`.

**Comment by rlepigre** [comment]
Note that a workaround is to write an empty file to disk, that is immediately deleted after the toplevel has been started. However, that is a bit annoying to do, and I think it would be easy to find the relevant stanza, and extend the list of source files with the one given as argument (or something like that).

**Comment by rlepigre** [comment]
I see, you're right that it is non-trivial, at least in complicated setups. Maybe we should only support this if there is a single `coq.theory` stanza with `(modules :standard)`? I think that should take care of 99% of the pain some people have, and we could give an error saying that the stanza for the input file cannot be determined unambiguously otherwise.

**Comment by ejgallego** [comment]
> see, you're right that it is non-trivial, at least in complicated setups. Maybe we should only support this if there is a single `coq.theory` stanza with `(modules :standard)`? I think that should take care of 99% of the pain some people have, and we could give an error saying that the stanza for the input file cannot be determined unambiguously otherwise.

I'm afraid that the code to do that could be even more complex than just injecting a file in the filesystem, which happens at an earlier phase (VFS build).

There is a little bit of a hack that maybe could work without too much effort: if the file doesn't exist, create it, then just run dune.

You could do that in the code for `dune coq top` or just as a shell script wrapper to that command.

**Comment by rlepigre** [comment]
> There is a little bit of a hack that maybe could work without too much effort: if the file doesn't exist, create it, then just run dune.

That's exactly what some of us have been doing, but I was hoping to move towards not having to hack around problems with a wrapper script. :smile:

**Comment by Alizter** [comment]
Coming back to this, is the suggestion for `dune coq top` to create the file if it doesn't exist?

**Comment by rlepigre** [comment]
> Coming back to this, is the suggestion for `dune coq top` to create the file if it doesn't exist?

I think that is one possibility, yes.

**Comment by Alizter** [comment]
I'm a bit weary of doing something like that in case we have something like `dune coq top typo.v` where the user may have intended `typ.v`. On the other hand, a `--create-missing` argument might be a good compromise.

**Comment by rlepigre** [comment]
Yeah, I'm actually not entirely sold on the idea either. Maybe this whole use-case does not really matter, and the user can just save their file before running the top-level.

**Comment by ejgallego** [comment]
Indeed it seems to me that Dune requires the file to be present in the filesystem; I'm not sure it would work fine, but we could do as Ali proposes just having `dune coq top` create a rule for `new_file.v`, with empty contents, if the file doesn't exists.

It seems to me that guarding this with a command line option could be a good idea so `dune coq top foo.v` fails by default if the file is not there.
